multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/ApplicationDigestCalculator.java

```diff
@@ -0,0 +1,106 @@
+            return;
+        }
+        // TODO: backwards compatibility for one tact
+        if (ArchiveEntryExtractorUtil.containsDirectory(applicationArchiveContext.getModuleFileName(),
+                                                        applicationArchiveContext.getArchiveEntryWithStreamPositions())) {
+            fileService.consumeFileContent(applicationArchiveContext.getSpaceId(), applicationArchiveContext.getAppArchiveId(),
+                                           archiveStream -> calculateDigestFromDirectory(applicationArchiveContext, archiveStream));
+        } else {
+            ArchiveEntryWithStreamPositions archiveEntryWithStreamPositions = ArchiveEntryExtractorUtil.findEntry(applicationArchiveContext.getModuleFileName(),
+                                                                                                                  applicationArchiveContext.getArchiveEntryWithStreamPositions());
+            DigestCalculator applicationDigestCalculator = applicationArchiveContext.getDigestCalculator();
+            archiveEntryExtractor.processFileEntryContent(ImmutableFileEntryProperties.builder()
+                                                                                      .guid(applicationArchiveContext.getAppArchiveId())
+                                                                                      .name(archiveEntryWithStreamPositions.getName())
+                                                                                      .spaceGuid(applicationArchiveContext.getSpaceId())
+                                                                                      .maxFileSizeInBytes(applicationArchiveContext.getMaxSizeInBytes())
+                                                                                      .build(),
+                                                          archiveEntryWithStreamPositions, (bytesBuffer, bytesRead) -> {
+                                                              applicationArchiveContext.calculateCurrentSizeInBytes(bytesRead);
+                                                              applicationDigestCalculator.updateDigest(bytesBuffer, 0, bytesRead);
+                                                          });
+        }
+    }
+
+    private void calculateDigestFromDirectory(ApplicationArchiveContext applicationArchiveContext, InputStream archiveStream)
+        throws IOException {
+        try (ZipArchiveInputStream zipArchiveInputStream = new ZipArchiveInputStream(archiveStream)) {
+            String moduleFileName = applicationArchiveContext.getModuleFileName();
+            ZipEntry zipEntry = applicationArchiveIterator.getFirstZipEntry(applicationArchiveContext.getModuleFileName(),
+                                                                            zipArchiveInputStream);
+            do {
+                if (!zipEntry.isDirectory()) {
```

Comment:
what do we do if an entry is a directory?

Reply:
they do not contain actual information so it does not matter for the digest